src/adc.rs

```diff
@@ -0,0 +1,807 @@
+//! Analog-to-digital converter
+//!
+use core::marker::PhantomData;
+
+// NOTE: embedded_hal's Channel is not suitable for this rank + channel style ADC.
```

Comment:
Could you elaborate on this? I believe that GD32V ADC should be "the same" as the one used in STM32F103, and the stm32f1xx-hal ADC driver implements embedded-hal traits.

Reply:
They are not the same.

GD32VF103's channel number is actually the `rank` number when configuring, not the `Channel::channel()` number.

`Channel::channel()` in `embedded_hal` is a method bonded to type, not the self-object. So one can not save the real channel number in a wrapped "Channel" type.

Reply:
The `OneShot` abstraction is not suitable for GD32VF103. So I left it unimplemented.

Reply:
What do you mean by rank ? I mean there is 18 multiplexed channels like on stm32f103.

From my understanding, you can implement The OneShot mode as describe in documentation in section "Single conversion mode" with the sequence configured with only the desired channel.

Reply:
`rank` is the parameter name defined in C SDK.
`Single conversion mode` might help, will try to implement it. 😁